storage: improve `ComputeStats` performance

erikgrinaker opened the pull request:
This patch omits a `sort.Search()` call during `ComputeStats()` when
possible, which essentially recovers the remaining no-range-key
regression since 22.1.

```
name                                                      old time/op   new time/op   delta
MVCCComputeStats_Pebble/valueSize=32/numRangeKeys=0-24      156ms ± 1%    145ms ± 1%   -7.28%  (p=0.000 n=10+10)
MVCCComputeStats_Pebble/valueSize=32/numRangeKeys=1-24      190ms ± 1%    178ms ± 1%   -6.24%  (p=0.000 n=10+10)
MVCCComputeStats_Pebble/valueSize=32/numRangeKeys=100-24    210ms ± 1%    187ms ± 1%  -10.89%  (p=0.000 n=10+10)
```

Resolves #84544.

Release justification: bug fixes and low-risk updates to new functionality

Release note: None

erikgrinaker committed:
storage: improve `ComputeStats` performance

This patch omits a `sort.Search()` call during `ComputeStats()` when
possible, which essentially recovers the remaining no-range-key
regression since 22.1.

```
name                                                      old time/op   new time/op   delta
MVCCComputeStats_Pebble/valueSize=32/numRangeKeys=0-24      156ms ± 1%    145ms ± 1%   -7.28%  (p=0.000 n=10+10)
MVCCComputeStats_Pebble/valueSize=32/numRangeKeys=1-24      190ms ± 1%    178ms ± 1%   -6.24%  (p=0.000 n=10+10)
MVCCComputeStats_Pebble/valueSize=32/numRangeKeys=100-24    210ms ± 1%    187ms ± 1%  -10.89%  (p=0.000 n=10+10)
```

Release justification: bug fixes and low-risk updates to new functionality

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86515)
<!-- Reviewable:end -->

erikgrinaker committed:
storage: omit `sort.Search` in `MVCCRangeKeyStack.HasAbove/Below`

Release justification: bug fixes and low-risk updates to new functionality

Release note: None

aliher1911 reviewed (approved):
🚀

erikgrinaker commented:
TFTR!

bors r=aliher1911
